In [2]:
# Cargar las bibliotecas necesarias
import pandas as pd
import numpy as np

In [3]:
# Leer el archivo de datos descargado. Estación en Campo Experimental UniSon.
# Datos separados por espacios, ignorando los primeros 20 renglones y el último renglón.
# Se reasignan nombres a las columnas de la tabla de datos (Sin espacios, ni caracteres especiales). 
df0 = pd.read_csv("26174.txt", sep="\s+", skiprows=20, skipfooter=1, engine='python', 
                  names=['Fecha', 'Precip', 'Evap', 'Tmax', 'Tmin'])
# Reemplazar  los Nulos con NA (datos no disponibles)
df0 = df0.replace('Nulo','NA')
# Verificar
df0.head(), df0.tail()

(        Fecha Precip Evap  Tmax  Tmin
 0  14/10/1981      0   NA    28    NA
 1  15/10/1981      0  7.3    26    14
 2  16/10/1981      0  7.6    28  12.5
 3  17/10/1981      0  8.1    28    10
 4  18/10/1981      0  1.2  30.5   9.5,
            Fecha Precip Evap  Tmax  Tmin
 5439  26/09/1998      0   NA  35.5    NA
 5440  27/09/1998      0   NA  37.5  16.5
 5441  28/09/1998      0   NA  38.5    20
 5442  29/09/1998      0   NA  38.5    20
 5443  30/09/1998      0   NA  37.5    20)

In [4]:
# Ver que tipos de variables se han reconocido (Tipo Object no está definido)
df0.dtypes

Fecha     object
Precip    object
Evap      object
Tmax      object
Tmin      object
dtype: object

In [5]:
# Convertir Fecha a nueva variable tipo fecha NFecha (datetime64)
df0['NFecha'] =  pd.to_datetime(df0['Fecha'], format='%d/%m/%Y')
# Convertir las otras variables a flotantes (float64)
df0[['Precip','Evap','Tmax','Tmin']] = df0[['Precip','Evap','Tmax','Tmin']] \
.apply(pd.to_numeric, errors='coerce')
# Eliminar columna Fecha que ya no se utilzará
df0 = df0.drop('Fecha', 1)
# Verificar
df0.dtypes

Precip           float64
Evap             float64
Tmax             float64
Tmin             float64
NFecha    datetime64[ns]
dtype: object

** Resumen de los datos: ** <br>



In [6]:
# Explorar un resumen de los datos
df0.describe()

Precip         Evap         Tmax         Tmin
count  5182.000000  3104.000000  4974.000000  4959.000000
mean      0.972096     6.432571    32.110032    13.363783
std       5.666403     3.219988     6.949653     7.035267
min       0.000000     0.100000     0.000000    -3.500000
25%       0.000000     3.800000    26.625000     7.500000
50%       0.000000     6.100000    33.000000    12.000000
75%       0.000000     8.800000    37.500000    19.500000
max     119.300000    16.500000    49.000000    29.000000

In [7]:
# Crear columnas con Año y Mes extraídos de la fecha 
df0['Año'] = df0['NFecha'].dt.year
df0['Mes'] = df0['NFecha'].dt.month
df0.head(),df0.tail()

(   Precip  Evap  Tmax  Tmin     NFecha   Año  Mes
 0     0.0   NaN  28.0   NaN 1981-10-14  1981   10
 1     0.0   7.3  26.0  14.0 1981-10-15  1981   10
 2     0.0   7.6  28.0  12.5 1981-10-16  1981   10
 3     0.0   8.1  28.0  10.0 1981-10-17  1981   10
 4     0.0   1.2  30.5   9.5 1981-10-18  1981   10,
       Precip  Evap  Tmax  Tmin     NFecha   Año  Mes
 5439     0.0   NaN  35.5   NaN 1998-09-26  1998    9
 5440     0.0   NaN  37.5  16.5 1998-09-27  1998    9
 5441     0.0   NaN  38.5  20.0 1998-09-28  1998    9
 5442     0.0   NaN  38.5  20.0 1998-09-29  1998    9
 5443     0.0   NaN  37.5  20.0 1998-09-30  1998    9)

In [8]:
# Número de años distintos df0['Año'].unique(),
NumA = len(df0['Año'].unique())
print(NumA, "años de datos")

18 años de datos


In [9]:
# Calcular la precipitación promedio de un mes (Sumar Precip y dividir entre número de Años)
# Mes a mes ... a pie
PrecipEne = df0[(df0['Mes'] == 1)]['Precip'].sum()/NumA
PrecipFeb = df0[(df0['Mes'] == 2)]['Precip'].sum()/NumA
print("Enero:",PrecipEne, "mm")
print("Febrero:",PrecipFeb, "mm")

Enero: 15.4388888889 mm
Febrero: 18.5222222222 mm


In [11]:
# Para encontrar los meses más lluviosos
# O hacer un loop para recorrer cada mes, recordando que Python cuenta desde 0,1,2,...
# Calcular la suma de precipitación y dividir entre el número de años
total=0.0
for i in range(12):
    PrecipMensual = df0['Precip'][df0['Mes']==[i+1]].sum()/NumA
    total=total+PrecipMensual
    print("Mes", i+1,":", np.round(PrecipMensual, decimals=2), "mm", ", Acumulada:", 
          np.round(total, decimals=2), "mm")

Mes 1 : 15.44 mm , Acumulada: 15.44 mm
Mes 2 : 18.52 mm , Acumulada: 33.96 mm
Mes 3 : 4.98 mm , Acumulada: 38.94 mm
Mes 4 : 2.43 mm , Acumulada: 41.37 mm
Mes 5 : 3.75 mm , Acumulada: 45.12 mm
Mes 6 : 1.88 mm , Acumulada: 47.0 mm
Mes 7 : 51.84 mm , Acumulada: 98.84 mm
Mes 8 : 78.32 mm , Acumulada: 177.16 mm
Mes 9 : 42.91 mm , Acumulada: 220.07 mm
Mes 10 : 6.91 mm , Acumulada: 226.98 mm
Mes 11 : 18.04 mm , Acumulada: 245.02 mm
Mes 12 : 34.83 mm , Acumulada: 279.86 mm


**Meses más lluviosos:**
Julio 51.84mm, Agosto 78.32mm y Septiembre 42.91mm

In [12]:
# Calcular la Tmax promedio de cada mes. Sumar las Tmax y 
# dividir entre el número total de datos válidos para ese mes determinado (función .count())
TmaxPromEne = df0[df0['Mes']==1]['Tmax'].sum()/df0[df0['Mes'] ==1]['Tmax'].count()
TmaxPromFeb = df0[df0['Mes']==2]['Tmax'].sum()/df0[df0['Mes'] ==2]['Tmax'].count()
print("Tmax promedio Ene:", np.round(TmaxPromEne, decimals=2), "ºC")
print("Tmax promedio Feb:", np.round(TmaxPromFeb, decimals=2), "ºC")

Tmax promedio Ene: 24.19 ºC
Tmax promedio Feb: 25.76 ºC


In [13]:
# Calcular la temperatura promedio Tmin (meses fríos) y Tmax (meses cálidos)

for i in range(12):
    TminPromMensual = df0[df0['Mes']==i+1]['Tmin'].sum()/df0[df0['Mes'] ==i+1]['Tmin'].count()
    print("Tmin Mes", i+1,":", np.round(TminPromMensual, decimals=2), "ºC")
  
print(" ")
    
for i in range(12):
    TmaxPromMensual = df0[df0['Mes']==i+1]['Tmax'].sum()/df0[df0['Mes'] ==i+1]['Tmax'].count()
    print("Tmax Mes", i+1,":", np.round(TmaxPromMensual, decimals=2), "ºC")


Tmin Mes 1 : 5.98 ºC
Tmin Mes 2 : 7.1 ºC
Tmin Mes 3 : 8.47 ºC
Tmin Mes 4 : 11.16 ºC
Tmin Mes 5 : 14.28 ºC
Tmin Mes 6 : 18.72 ºC
Tmin Mes 7 : 22.88 ºC
Tmin Mes 8 : 23.52 ºC
Tmin Mes 9 : 21.7 ºC
Tmin Mes 10 : 15.51 ºC
Tmin Mes 11 : 9.29 ºC
Tmin Mes 12 : 6.38 ºC
 
Tmax Mes 1 : 24.19 ºC
Tmax Mes 2 : 25.76 ºC
Tmax Mes 3 : 28.59 ºC
Tmax Mes 4 : 32.62 ºC
Tmax Mes 5 : 35.92 ºC
Tmax Mes 6 : 39.74 ºC
Tmax Mes 7 : 38.86 ºC
Tmax Mes 8 : 38.51 ºC
Tmax Mes 9 : 37.79 ºC
Tmax Mes 10 : 34.64 ºC
Tmax Mes 11 : 28.78 ºC
Tmax Mes 12 : 23.85 ºC


**Meses fríos:** Noviembre, Diciembre, Enero, Febrero y Marzo (Tmin < 10ºC) <br>
**Meses cálidos:** Mayo, Junio, Julio, Agosto, Septiembre (Tmax > 35ºC)

In [14]:
# Años húmedos y años secos
# Años incompletos 1981, 1998 - se excluyen
for i in range(1982,1997):
    PrecipAnual = df0['Precip'][df0['Año']==[i+1]].sum()
    print("Año", i+1,":", PrecipAnual, "mm")

Año 1983 : 445.9 mm
Año 1984 : 457.2 mm
Año 1985 : 261.1 mm
Año 1986 : 302.1 mm
Año 1987 : 135.6 mm
Año 1988 : 287.0 mm
Año 1989 : 99.8 mm
Año 1990 : 528.2 mm
Año 1991 : 422.5 mm
Año 1992 : 613.6 mm
Año 1993 : 329.9 mm
Año 1994 : 458.0 mm
Año 1995 : 208.0 mm
Año 1996 : 135.1 mm
Año 1997 : 123.5 mm


In [15]:
# Número de datos faltantes por año
for i in range(1982,1997):
    NumDatos= df0['Precip'][df0['Año']==[i+1]].count()
    print("Año:", i+1, "Num Datos", NumDatos, ", ", np.round(NumDatos/365.0, decimals=2), "%")
    

Año: 1983 Num Datos 305 ,  0.84 %
Año: 1984 Num Datos 338 ,  0.93 %
Año: 1985 Num Datos 364 ,  1.0 %
Año: 1986 Num Datos 271 ,  0.74 %
Año: 1987 Num Datos 356 ,  0.98 %
Año: 1988 Num Datos 340 ,  0.93 %
Año: 1989 Num Datos 244 ,  0.67 %
Año: 1990 Num Datos 299 ,  0.82 %
Año: 1991 Num Datos 356 ,  0.98 %
Año: 1992 Num Datos 322 ,  0.88 %
Año: 1993 Num Datos 304 ,  0.83 %
Año: 1994 Num Datos 293 ,  0.8 %
Año: 1995 Num Datos 312 ,  0.85 %
Año: 1996 Num Datos 252 ,  0.69 %
Año: 1997 Num Datos 254 ,  0.7 %


**Año más húmedo:** 1992 (613 mm) (Existe sólo el 88% de datos anuales)
**Año muy seco:** 1989 (99.8 mm) (Existe sólo el 67% de datos anuales)